source/Mlos.Python/mlos/Optimizers/unit_tests/TestParetoFrontier.py

```diff
@@ -0,0 +1,287 @@
+            dimensions=[
+                ContinuousDimension(name='y1', min=0, max=10),
+                ContinuousDimension(name='y2', min=0, max=10)
+            ]
+        )
+
+        optimization_problem = OptimizationProblem(
+            parameter_space=parameter_space,
+            objective_space=objective_space,
+            objectives=[
+                Objective(name='y1', minimize=False),
+                Objective(name='y2', minimize=False)
+            ]
+        )
+
+        num_rows = 100000
+        random_params_df = parameter_space.random_dataframe(num_rows)
+        random_objectives_df = objective_space.random_dataframe(num_rows)
+
+        pareto_df = ParetoFrontier.compute_pareto(
+            optimization_problem=optimization_problem,
+            objectives_df=random_objectives_df
+        )
+
+        non_pareto_index = random_objectives_df.index.difference(pareto_df.index)
+        for i, row in pareto_df.iterrows():
+            # Now let's make sure that no point in pareto is dominated by any non-pareto point.
+            #
+            assert (random_objectives_df.loc[non_pareto_index] < row).any(axis=1).sum() == len(non_pareto_index)
+
+            # Let's also make sure that no point on the pareto is dominated by any other point there.
+            #
+            other_rows = pareto_df.index.difference([i])
+            assert (pareto_df.loc[other_rows] > row).all(axis=1).sum() == 0
+
+
+    @pytest.mark.parametrize("minimize", ["all", "none", "some"])
+    @pytest.mark.parametrize("num_output_dimensions", [2, 10])
+    @pytest.mark.parametrize("num_points", [100, 1000])
+    def test_hyperspheres(self, minimize, num_output_dimensions, num_points):
+        """Uses a hypersphere to validate that ParetoFrontier can correctly identify pareto-optimal points.
+
+
+        The idea is that we want to find a pareto frontier that optimizes the cartesian coordinates of points defined using random
+        spherical coordinates.
+
+        By setting the radius of some of the points to the radius of the hypersphere, we guarantee that they are non-dominated.
+        Such points must appear on the pareto frontier, though it's quite possible that other non-dominated points from the interior
+        of the sphere could appear as well. The intuition in 2D is that we can draw a secant between two neighboring pareto efficient
+        points on the perimeter. Any point that is between that secant and the perimeter is not dominated and would thus be pareto
+        efficient as well. (Actually even more points are pareto efficient, but this subset is easiest to explain in text).
+
+
+        We want to test scenarios where:
+            1) all objectives are maximized,
+            2) all objectives are minimized,
+            3) some objectives are maximized and some are minimized.
+
+        We want to be able to do that for an arbitrary number of dimensions so as to extract maximum coverage from this simple test.

```

Comment:
Just as a side note, in higher dimensions, nearly all the points will be on the pareto frontier I think.

Reply:
Yeah.